code/controllers/subsystem/minimaps.dm

```diff
@@ -0,0 +1,281 @@
+
+/**
+ * Holder datum to ease updating of atoms to update
+ */
+/datum/minimap_updator
+	/// Atom to update with the overlays
+	var/atom/minimap
+	///Target zlevel we want to be updating to
+	var/ztarget = 0
+
+/datum/minimap_updator/New(minimap, ztarget)
+	..()
+	src.minimap = minimap
+	src.ztarget = ztarget
+
+/**
+ * Adds an atom we want to track with blips to the subsystem
+ * Arguments:
+ * * target: atom we want to track
+ * * zlevel: zlevel we want thhis atom to be tracked for
+ * * hud_flags: tracked HUDs we want this atom to be displayed on
+ * * iconstate: iconstate for the blip we want to be used for this tracked atom
+ * * icon: icon file we want to use for this blip, 'icons/UI_icons/map_blips.dmi' by efault
+ */
+/datum/controller/subsystem/minimaps/proc/add_marker(atom/target, zlevel, hud_flags = NONE, iconstate, icon = 'icons/UI_icons/map_blips.dmi')
+	if(!isatom(target) || !zlevel || !hud_flags || !iconstate || !icon)
+		CRASH("Invalid marker added to subsystem")
+	if(!initialized)
+		earlyadds += CALLBACK(src, .proc/add_marker, target, zlevel, hud_flags, iconstate, icon)
+		return
+	var/image/blip = image(icon, iconstate, pixel_x = MINIMAP_PIXEL_FROM_WORLD(target.x), pixel_y = MINIMAP_PIXEL_FROM_WORLD(target.y))
+	images_by_source[target] = blip
+	if(hud_flags & MINIMAP_FLAG_XENO)
+		minimaps_by_z["[zlevel]"].xeno_images[target] = blip
+		minimaps_by_z["[zlevel]"].xeno_images_raw += blip
+	if(hud_flags & MINIMAP_FLAG_MARINE)
+		minimaps_by_z["[zlevel]"].marine_images[target] = blip
+		minimaps_by_z["[zlevel]"].marine_images_raw += blip
+	if(hud_flags & MINIMAP_FLAG_ALPHA)
+		minimaps_by_z["[zlevel]"].alpha_images[target] = blip
+		minimaps_by_z["[zlevel]"].alpha_images_raw += blip
+	if(hud_flags & MINIMAP_FLAG_BRAVO)
+		minimaps_by_z["[zlevel]"].bravo_images[target] = blip
+		minimaps_by_z["[zlevel]"].bravo_images_raw += blip
+	if(hud_flags & MINIMAP_FLAG_CHARLIE)
+		minimaps_by_z["[zlevel]"].charlie_images[target] = blip
+		minimaps_by_z["[zlevel]"].charlie_images_raw += blip
+	if(hud_flags & MINIMAP_FLAG_DELTA)
+		minimaps_by_z["[zlevel]"].delta_images[target] = blip
+		minimaps_by_z["[zlevel]"].delta_images_raw += blip
+	if(ismovableatom(target))
+		RegisterSignal(target, COMSIG_MOVABLE_Z_CHANGED, .proc/on_z_change)
+		RegisterSignal(target, COMSIG_MOVABLE_MOVED, .proc/on_move)
+	removal_cbs[target] = CALLBACK(src, .proc/removeimage, blip, target)
+	RegisterSignal(target, COMSIG_PARENT_QDELETING, .proc/remove_marker)
+
+/**
+ * removes an image from raw tracked lists, invoked by callback
+ */
+/datum/controller/subsystem/minimaps/proc/removeimage(image/blip, atom/target)
+	minimaps_by_z["[target.z]"].xeno_images_raw -= blip
+	minimaps_by_z["[target.z]"].marine_images_raw -= blip
+	minimaps_by_z["[target.z]"].alpha_images_raw -= blip
+	minimaps_by_z["[target.z]"].bravo_images_raw -= blip
+	minimaps_by_z["[target.z]"].charlie_images_raw -= blip
+	minimaps_by_z["[target.z]"].delta_images_raw -= blip
+	removal_cbs -= target
```

Comment:
How efficient is finding an item in a list? Why not asking for hud flags here? since removeimage is only in a callback from addmarker where you have access to the flags

Reply:
I'd prefer this to more if() spam than necessary

Reply:
fixed meanwhile